Primitives vs object (#28440)
TheOdinProject/curriculum: foundations/javascript_basics/object_basics.md

Reviewer: I'd also like for there to be a [note box](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md#note-boxes) at the end of the new section just giving a quick example of how reassigning the original object variable won't magically reassign any other variables that reference the same object.

e.g.
```js
let foo = { foo: "foo" };
let fooObjRef = obj;
foo = "foo";

console.log(obj); // "foo"
console.log(fooObjRef); // { foo: "foo" }
```
But a better example of course :P

Author: While changing the object we have a reference to will change that object for all other variables that reference it, changing what we refer to won’t affect what the other variables refer to. This example should clarify it:
```js
let dog = { voice: "bark" };
// cat will have reference to the same object as dog variable
let cat = dog;

// changing the reference itself, cat now reference something else
cat = ["meow"];

console.log(dog); // { voice: "bark" }
console.log(cat); // ["meow"]

// this is also true when trying to change the reference within a function
function tryToReasign(obj) {
  obj = { dont: "work" };
}
tryToReasign(dog);
// dog won't change, we changed what obj refers to
// we didn't changed the reference of dog
console.log(dog); // { voice: "bark" }
```

Do you think that's good enough or a better example/phrasing is needed?

Reviewer: I don't think we need the function example. The first example is sufficient IMO. I'd prefer it if it demonstrated reassigning the original variable though, as that seems to trip people up more than the order you demonstrated for some reason :shrug: 

e.g.
```js
let animal = { species: "dog" };
// dog now references the same object as the animal variable
let dog = animal;

// reassigning animal so it now references a different object
animal = { species: "cat" };

// even though the animal variable now references something different,
// the dog variable still references the original object
console.log(animal); // { species: "cat" }
console.log(dog); // { species: "dog" }
```

Reviewer: Would also love an extra knowledge check question or two (whatever is appropriate) for some of the new stuff.

Author: The main "usage" of references is passing information to functions (as far as I see it), so I think that if a learner can answer this question they're good to go:
> How will a primitive and an object data type variables will effected by changes when passed to a function?

Unless you have any other suggestion (or rephrasing of the proposed question) :-)

Reviewer: Hmmm, what do you think about
> - How do primitives and object types differ when you assign the to other variables or pass them into functions?

1. This JavaScript.info [article on objects](https://javascript.info/object) is the best place to get started.
1. The [MDN tutorial on objects](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Basics) isn't bad either, so check it out if you need another take on the subject.

#### Differences between objects and primitives

Earlier in the curriculum you've learned about [primitive data types](https://www.theodinproject.com/lessons/foundations-data-types-and-conditionals). Now you've seen the object data type, the main difference between the two is that primitives can contain only a single thing (string, number, etc). Objects are used to store collection of data and more complex entities.
Besides the formal differences, there are also some technical differences, which affect how we use each data type.

> When you define a primitive variable, it will contain a copy of the information provided to it:

Author: Regarding the formatting I'll change it no problem.
Regarding this:
> Now you've seen the object data type
Maybe I phrased it incorrectly, but I was trying to say that previously the learner have seen the primitive types (the link), and **now** (a section above) he have seen the object type. Regarding the data type maybe something like this would work?:
> Now you've seen the object **data type**, which includes but not limited to, objects (`{key: value}`), arrays and functions. The main difference between the two is that primitives can contain only a single thing (string, number, etc). Objects are used to store collection of data and more complex entities.

```javascript

> const data = 42;

Author: I actually was thinking it may mislead learners for same reason, and if it's not just me then I agree it should be changed to reduce confusion, even tho it's the "correct" way to write this code which is why I decided to go for const in the end.

> const data = 42;

Reviewer: Yeah, I'd say `let` is the way to go. I've used `const` in several examples before and when asked for the reasons, people have answer "it won't change because it's a const variable"

```javascript
const element = document.getElementById("container");
element.style.backgroundColor = "red";
```

We're changing the variable we declared (element), yet the changes affect the corresponding node in the DOM.
Why does it happen?
That's because the node we have in our code is a **reference** to the same node that our DOM uses. If that wasn't a reference, but a copy like primitive data types behave, our changes would have **no** effect! Because the changes would be made to the local copy we have.

This behavior is also something to consider when we pass arguments to a function, let's take the following functions for example:

```javascript
function increaseCounterObject(objectCounter) {
  objectCounter.counter += 1;
}

function increaseCounterPrimitive(primitiveCounter) {
  primitiveCounter += 1;
}

const object = { counter: 0 };
let primitive = 0;

increaseCounterObject(object);
increaseCounterPrimitive(primitive);
```

Take a moment to try to guess what will happen to `object` and what will happen to `primitive`, after we make the function calls.

If you answered that the object counter would increase by 1, and the primitive counter wouldn't change, you're correct.

> Remember that `objectCounter` is passed by reference, while `primitiveCounter` is passed by copy.

Author: Agree, only a little fix, can commit-push it but not sure how to edit it on GH itself.
> Remember that `objectCounter` contains a reference to the same object as the `object` variable we gave it, while `primitiveCounter` only contains a copy of the primitive value.

> Remember that `objectCounter` is passed by reference, while `primitiveCounter` is passed by copy.

Reviewer: `git fetch origin primitives-vs-object && git merge origin/primitives-vs-object` to pull the commits you've made here to your cloned branch. Then you're in sync and can just make changes locally and commit and push.

1. This JavaScript.info [article on objects](https://javascript.info/object) is the best place to get started.
1. The [MDN tutorial on objects](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Basics) isn't bad either, so check it out if you need another take on the subject.

### Differences between objects and primitives

> Earlier in the curriculum you've learned about [primitive data types](https://www.theodinproject.com/lessons/foundations-data-types-and-conditionals). Now that you've seen the object data type, which includes but is not limited to, objects ({key: value}), arrays, and functions. The main difference between the two is that primitives can contain only a single thing (string, number, etc). Objects data types are used to store collection of data and more complex entities.

Reviewer: ```suggestion
Earlier in the curriculum you've learned about [primitive data types](https://www.theodinproject.com/lessons/foundations-data-types-and-conditionals). Now that you've seen the object data type, which includes but is not limited to, objects ({key: value}), arrays, and functions. The main difference between the two is that primitives can contain only a single thing (string, number, etc). Objects data types are used to store a collection of data and more complex entities.
```

> Besides the formal differences, there are also some technical differences, which affect how we use each data type.
> When you define a primitive variable, it will contain a copy of the information provided to it:

Reviewer: ```suggestion
Besides the formal differences, there are also some technical differences which affect how we use each data type. When you define a primitive variable, it will contain a copy of the information provided to it:
```

```javascript
let data = 42;
// dataCopy will store a copy of what data contains, so a copy of 42
let dataCopy = data;

// which means that making changes to dataCopy won't affect data
dataCopy = 43;

console.log(data); // 42
console.log(dataCopy); // 43
```

On the other hand, when you define an object variable, it will contain a *reference* to the object provided to it:

```javascript
// obj contains a reference to the object we defined on the right side
const obj = { data: 42 };
// objCopy will contain a reference to the object referenced by obj
const objCopy = obj;

// making changes to objCopy will make changes to the object that it refers to
objCopy.data = 43;

console.log(obj); // { data: 43 }
console.log(objCopy); // { data: 43 }
```

> This behavior isn't new to you. In your last project, you made changes to the cells in the Etch-A-Sketch grid by using references.

Reviewer: ```suggestion
This behavior isn't new to you. In your last project, you made changes to the cells in the Etch-A-Sketch grid by using references.

```

> This behavior isn't new to you. In your last project, you made changes to the cells in the Etch-A-Sketch grid by using references.
> Let's take this code snippet as an example:

Reviewer: ```suggestion
This behavior isn't new to you. In your last project, you made changes to the cells in the Etch-A-Sketch grid by using references. Let's take this code snippet as an example:
```

```javascript
const element = document.querySelector("#container");
element.style.backgroundColor = "red";
```

> We're mutating the variable we declared (`element`), yet the changes affect the corresponding node in the DOM.
> Why does it happen?
> That's because the node we have in our code is a **reference** to the same node that our DOM uses. If that wasn't a reference, but a copy like primitive data types behave, our changes would have **no** effect! Because the changes would be made to the local copy we have.

Reviewer: Single line breaks don't do anything different, so we keep these sentences on the same line (unless you want to separate into paragraphs, in which case separate with a single blank line).
```suggestion
We're mutating the variable we declared (`element`), yet the changes affect the corresponding node in the DOM. Why does it happen? That's because the node we have in our code is a **reference** to the same node that our DOM uses. If that wasn't a reference, but a copy like primitive data types behave, our changes would have **no** effect! Because the changes would be made to the local copy we have.
```

> This behavior is also something to consider when we pass arguments to a function, let's take the following functions for example:

Reviewer: Nit
```suggestion
This behavior is also something to consider when we pass arguments to a function. Let's take the following functions for example:
```

```javascript
function increaseCounterObject(objectCounter) {
  objectCounter.counter += 1;
}

function increaseCounterPrimitive(primitiveCounter) {
  primitiveCounter += 1;
}

const object = { counter: 0 };
let primitive = 0;

increaseCounterObject(object);
increaseCounterPrimitive(primitive);
```

Take a moment and guess what will happen to `object` and what will happen to `primitive` after we make the function calls.

> If you answered that the object counter would increase by 1, and the primitive counter wouldn't change, you're correct.
> Remember that `objectCounter` contains a reference to the same object as the `object` variable we gave it, while `primitiveCounter` contains only a copy of the primitive value only.

Reviewer: Fixing line formatting and small nit
```suggestion
If you answered that the object counter would increase by 1, and the primitive counter wouldn't change, you're correct. Remember that the parameter `objectCounter` contains a *reference* to the same object as the `object` variable we gave it, while `primitiveCounter` contains only a copy of the primitive value only.
```

<div class="lesson-note" markdown="1">

#### Reassigning object data type variables

> While changing the object we have a reference to will change that object for all other variables that reference it, changing what we refer to won’t affect what the other variables refer to. This example should clarify it:

Reviewer: Again, preferring the explicit mutate/reassign terminology than using "changing" twice with two different meanings in the same sentence.
```suggestion
While mutating the object we have a reference to will affect all other variables that reference it, reassigning what a variable refers to won’t affect what the other variables refer to. For example:
```

- [What is the difference between objects and arrays?](https://javascript.info/object#summary)
- [How do you access object properties?](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Basics#bracket_notation)

> - [How do primitives and object types differ when you assign the to other variables or pass them into functions?](https://www.theodinproject.com/lessons/foundations-object-basics#differences-between-objects-and-primitives)

Reviewer: ```suggestion
- [How do primitives and object types differ when you assign them to other variables, or pass them into functions?](https://www.theodinproject.com/lessons/foundations-object-basics#differences-between-objects-and-primitives)
```
